Source/JavaScriptCore/yarr/YarrJIT.cpp

```diff
@@ -53,6 +53,64 @@ static constexpr bool verbose = false;
 JSC_ANNOTATE_JIT_OPERATION_RETURN(vmEntryToYarrJITAfter);
 #endif
 
+// We should pick the less frequently appearing character as a BM search's anchor to make BM search more and more efficient.
+// This class takes some samples from the passed subject string to put weight on characters so that we can pick an optimal one adaptively.
+class SubjectSampler {
+public:
+    static constexpr unsigned sampleSize = 128;
+
+    explicit SubjectSampler(CharSize charSize)
+        : m_is8Bit(charSize == CharSize::Char8)
+    {
+    }
+
+    int32_t frequency(UChar character) const
+    {
+        if (!m_size)
+            return 1;
+        return static_cast<int32_t>(m_samples[character & BoyerMooreBitmap::mapMask]) * 128 / m_size;
+    }
+
+    void sample(StringView string)
+    {
+        unsigned half = string.length() > sampleSize ? (string.length() - sampleSize) / 2 : 0;
+        unsigned count = 0;
+        if (string.is8Bit()) {
+            auto* characters8 = string.characters8();
+            for (unsigned i = half; i < string.length() && count < sampleSize; ++i) {
```

Comment:
You can compute this ahead:
```
    unsigned end = std::min(string.length(), hold + sampleSize);
```
And for the loop, just do:
```
    for (unsigned i = half; i < end; ++i) {
```
... and skip the `++count` too.

Reply:
Changed.